File: 剑指offer全解/52正则表达式匹配.py
# -*- coding:utf-8 -*-
class Solution:
    # s, pattern都是字符串
    def match(self, s, pattern):
        # write code here
        if(len(s) == 0 and len(pattern) == 0): return True
        if(len(s) != 0 and len(pattern) == 0): return False
        
        if(len(pattern) > 1 and pattern[1] == '*'):
            if s and (pattern[0] == '.' or s[0] == pattern[0]):
                return self.match(s[1:], pattern) or self.match(s, pattern[2:])
            else:
                return self.match(s, pattern[2:])
        elif s and (pattern[0] == '.' or s[0] == pattern[0]):
            return self.match(s[1:], pattern[1:])
        else:
            return False

File: 剑指offer全解/52正则表达式匹配.java
// DP

public class Solution {
    public boolean match(char[] str, char[] pattern)
    {
        if(str == null || pattern == null) return false;
        if(str.length == 0 && pattern.length == 0) return true;
        if(str.length != 0 && pattern.length == 0) return false;
        
        boolean [][] dp = new boolean[str.length+1][pattern.length+1];
        dp[str.length][pattern.length] = true;
        for(int i = str.length; i >= 0; i--) {
            for(int j = pattern.length-1; j >= 0; j--) {
                if(j < pattern.length - 1 && pattern[j+1] == '*') {
                    if(i < str.length && (str[i] == pattern[j] || pattern[j] == '.')) {
                        dp[i][j] = dp[i][j+2] || dp[i+1][j];
                    } else {
                        dp[i][j] = dp[i][j+2];
                    }
                } else {
                    if(i != str.length && (str[i] == pattern[j] || pattern[j] == '.')) {
                        dp[i][j] = dp[i+1][j+1];
                    }
                }
            }
        }
        return dp[0][0];
    }
}